tidy3d/components/mode/solver.py

```diff
@@ -1089,6 +1112,248 @@ def mode_plane_contain_good_conductor(material_response) -> bool:
+        E: np.ndarray,
+        H: np.ndarray,
+        dl_primal: np.ndarray,
+        dl_dual: np.ndarray,
+        tol: float,
+    ) -> list[set[int]]:
+        """Inspects the biorthogonality condition between modes to filter sets of degenerate modes.
+
+        For each set of modes identified as degenerate by neff similarity, this function computes
+        the overlap matrix and uses connected components to find which modes are truly degenerate
+        (have significant mutual overlap).
+
+        Returns only groups with >= 2 modes that have significant overlap.
+        """
+        import scipy.sparse as sp
+
+        filtered_degenerate_sets = []
+
+        for degenerate_group in degenerate_sets:
+            # Convert set to list for indexing
+            mode_indices = sorted(degenerate_group)
+
+            # S is a symmetric overlap matrix of the degenerate modes
+            S = EigSolver._outer_dot(E, H, dl_primal, dl_dual, mode_indices)
+
+            # Create adjacency matrix
+            # True if overlap is significant, False if orthogonal
+            # Zero out diagonal so a mode doesn't connect to itself
+            adjacency = (np.abs(S) > tol).astype(int)
+            np.fill_diagonal(adjacency, 0)
+
+            # Find connected components
+            # labels: array indicating which group each mode belongs to
+            _, labels = sp.csgraph.connected_components(sp.csr_matrix(adjacency), directed=False)
+
+            # Group modes by their connected component label
+            groups = {}
+            for label_idx, group_id in enumerate(labels):
+                mode_index = mode_indices[label_idx]
+                if group_id not in groups:
+                    groups[group_id] = []
+                groups[group_id].append(mode_index)
+
+            # Convert groups dict to list of sets, keeping only groups with >= 2 modes
+            for group_modes in groups.values():
+                if len(group_modes) >= 2:
+                    filtered_degenerate_sets.append(set(group_modes))
+
+        return filtered_degenerate_sets
+
+    @staticmethod
+    def _dot(
+        E: np.ndarray,
+        H: np.ndarray,
+        dl_primal: np.ndarray,
+        dl_dual: np.ndarray,
+        mode_1: int,
+        mode_2: int,
+    ) -> complex:
+        """Dot product based on the bi-orthogonality relationship between E and H."""
+        # Extract field components
+        Ex = E[0, ...]
+        Ey = E[1, ...]
+        Hx = H[0, ...]
+        Hy = H[1, ...]
+
+        # Make the differential area elements
+        Ex_Hy_dS = np.outer(dl_primal[0], dl_dual[1])
+        Ey_Hx_dS = np.outer(dl_dual[0], dl_primal[1])
+
+        term1 = Ex[..., mode_1] * Hy[..., mode_2] + Ex[..., mode_2] * Hy[..., mode_1]
+        term1 *= Ex_Hy_dS[..., np.newaxis]
+        term2 = Ey[..., mode_1] * Hx[..., mode_2] + Ey[..., mode_2] * Hx[..., mode_1]
+        term2 *= Ey_Hx_dS[..., np.newaxis]
+        return np.sum(term1 - term2)
```

Comment:
I think we want a factor of 1/2 here for consistency with the usage elsewhere in tidy3d (e.g. outer_dot with conj=False)